### PR TITLE
JS: use inheritance in `js/mixed-static-instance-this-access`

### PR DESCRIPTION
**Pull request by esben-semmle**
This change makes `js/mixed-static-instance-this-access` use the inheritance hierarchy. This can both add and remove results, as evidenced by the added tests.

The motivation for this change comes from these false positives: <https://lgtm.com/projects/g/Tencent/omi/alerts/?mode=tree&ruleFocus=2154240232>

I think we should keep the introduced `getASuperClassMethod` out of `ClassDefinition` for now. We can implement a precise and scalable member resolution when we have more use cases.

The [evaluation](https://git.semmle.com/esben/dist-compare-reports/tree/js/mixed-static-intance-this-access-inheritance_1542733754836) shows that performance is reasonable.

### CODE REVIEW

File: javascript/ql/src/Declarations/MixedStaticInstanceThisAccess.ql [hunk 1]
```diff
@@ -10,15 +10,20 @@
 */
 import javascript
 
+MethodDefinition getASuperClassMethod(ClassDefinition base, string name) {
+  result.getDeclaringClass() = base and
```
Comment (xiemaisi) [inline javascript/ql/src/Declarations/MixedStaticInstanceThisAccess.ql]:
```suggestion
  result = base.getMethod(name)
```

File: javascript/ql/src/Declarations/MixedStaticInstanceThisAccess.ql [hunk 2]
```diff
@@ -10,15 +10,20 @@
...
+  result.getName() = name
```
Comment (xiemaisi) [inline javascript/ql/src/Declarations/MixedStaticInstanceThisAccess.ql]:
```suggestion
```

File: javascript/ql/src/Declarations/MixedStaticInstanceThisAccess.ql [hunk 3]
```diff
@@ -10,15 +10,20 @@
...
+  or
+  result = getASuperClassMethod(base.getSuperClass().analyze().getAValue().(AbstractClass).getClass(), name)
```
Comment (xiemaisi) [inline javascript/ql/src/Declarations/MixedStaticInstanceThisAccess.ql]: Can we pull out the `base.getSuperClass()...` bit into a member predicate of `ClassDefinition`?

File: javascript/ql/src/semmle/javascript/Classes.qll [hunk 4]
```diff
@@ -216,6 +216,13 @@ class ClassDefinition extends @classdefinition, ClassOrInterface, AST::ValueNode
     )
   }
 
+  /**
+   * Gets a direct super class, if it can be determined.
+   */
+  ClassDefinition getADirectSuperClass() {
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/Classes.qll]: Not a fan of the name. Could we work "definition" into the name to disambiguate it with the existing (and badly named) `getSuperClass()`? Also, while it's technically true that there could be more than one direct super class, this is one of those cases where allowing for that possibility in the name and description causes more confusion than it's worth.

File: javascript/ql/src/semmle/javascript/Classes.qll [hunk 5]
```diff
@@ -216,6 +216,13 @@ class ClassDefinition extends @classdefinition, ClassOrInterface, AST::ValueNode
     )
   }
 
+  /**
+   * Gets a direct super class definition, if it can be determined.
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/Classes.qll]:
```suggestion
   * Gets the definition of the super class of this class, if it can be determined.
```

File: javascript/ql/src/Declarations/MixedStaticInstanceThisAccess.ql [hunk 6]
```diff
@@ -10,15 +10,18 @@
 */
 import javascript
 
+MethodDefinition getASuperClassMethod(ClassDefinition base, string name) {
+  result = base.getMethod(name) or
+  result = getASuperClassMethod(base.getDirectSuperClassDefinition(), name)
```
Comment (xiemaisi) [inline javascript/ql/src/Declarations/MixedStaticInstanceThisAccess.ql]:
```suggestion
  result = getASuperClassMethod(base.getSuperClassDefinition(), name)
```

I don't think we need the "direct" anymore (Java, for example, calls the corresponding predicate `getASupertype`).

File: javascript/ql/src/Declarations/MixedStaticInstanceThisAccess.ql [hunk 7]
```diff
@@ -10,15 +10,18 @@
 */
 import javascript
 
+MethodDefinition getASuperClassMethod(ClassDefinition base, string name) {
```
Comment (xiemaisi) [inline javascript/ql/src/Declarations/MixedStaticInstanceThisAccess.ql]: The "super class" bit in the name is misleading. Taking a cue from Java, we could rephrase this as

```suggestion
/** Holds if `base` declares or inherits method `m` with the given `name`. */
predicate hasMethod(ClassDefinition base, string name, Method m) {
```

**Comment by esben-semmle** [comment]
Another round of fixes. I would like to squash before merging.

**Comment by esben-semmle** [comment]
Squashed.
